### DISCUSSION TITLE
Is it possible to create a GraphQLSchema from TypeDefinitionRegistry and GraphQLCodeRegistry?

### DISCUSSION THREAD

**Discussion by alf** [graphql-java/graphql-java#3758] [Q&A] [answered]
I'm struggling a bit with understanding the difference between the RuntimeWiring and the GraphQLCodeRegistry.

My current understanding is that the RuntimeWiring is just used to create a GraphQLCodeRegistry, but I suspect that's not completely correct.

The reason why I ask is that we're trying to gradually get rid of GraphQL Java Kickstart in favor of writing DataFetchers directly. It's quite easy to extract the GraphQLCodeRegistry from the GraphQLSchema that Kickstart creates for us, but I can't reach the RuntimeWiring.

Thus the question: Is there an easy way to create a GraphQLSchema by combining our schema files (SDL) with a GraphQLCodeRegistry, or must we go via a RuntimeWiring?

**Comment by bbakerman** [comment, marked as answer]
The runtime wiring is used to help create the "runtime aspects" of a schema - eg data fetchers and type resolvers

It is used when you create a schema from a text file, eg from Schema Definition Language (SDL)

https://www.graphql-java.com/documentation/schema#creating-a-schema-using-the-sdl

Nominally the runtime wiring is asked via a call back "do you have a data fetcher" field this field and if so it puts it into the `GraphQLCodeRegistry` for you

However you can supply in a `GraphQLCodeRegistry` to start with and if it has the DF in it for that field than it will use it

```
        // if they have already wired in a fetcher - then leave it alone
        FieldCoordinates coordinates = FieldCoordinates.coordinates(parentType.getName(), fieldDefinition.getName());
        if (!buildCtx.getCodeRegistry().hasDataFetcher(coordinates)) {
```

The reason we have both a RuntimeWiring and a `GraphQLCodeRegistry` is that is possible to create a schema programmatically and not via SDL

https://www.graphql-java.com/documentation/schema#creating-a-schema-programmatically        


To your specially question

> Is it possible to create a GraphQLSchema from TypeDefinitionRegistry and GraphQLCodeRegistry?

not really not - the SchemaGenerator does LOTS of work to take a simple parsed TypeDefinitionRegistry and turn it into a valid schema.  If you were to take a TypeDefinitionRegistry and GraphQLCodeRegistry and try to make a schema then you owuld be rewriting SchemaGenerator

I think its best to consider `GraphQLCodeRegistry` as a generated artifact of schema generation and not the place you "need to put DFs" upfront if you are using SDL

```
runtimeWiring.type("Droid", typeWiring -> typeWiring
                    .dataFetcher("friends", StarWarsData.getFriendsDataFetcher())
            )
```

really is the same as

```
FieldCoordinates coordinates = FieldCoordinates.coordinates("Droid", "friends");
coderegistryBuilder.dataFetcher(coordinates, StarWarsData.getFriendsDataFetcher()
```
